crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -0,0 +1,313 @@
+use runix::installable::{AttrPath, Installable};
+use runix::{RunJson, RunTyped};
+use serde_json::{json, Value};
+use thiserror::Error;
+
+use crate::flox::Flox;
+use crate::providers::git::{GitCommandProvider as Git, GitProvider};
+
+/// Publish state before analyzing
+///
+/// Prevents other actions to commence without analyzing the package first
+struct Empty;
+
+/// Publish state after collecting nix metadata
+///
+/// Json value (ideally a [flox_types::catalog::CatalogEntry],
+/// but that's currently broken on account of some flakerefs)
+#[derive(Debug, Deref, DerefMut)]
+struct NixAnalysis(Value);
+
+/// State for the publish algorihm
+///
+/// Transitions through typestates to ensure we don't invoke invalid operations
+pub struct Publish<'flox, State> {
+    flox: &'flox Flox,
+    /// The published _upstream_ source
+    publish_ref: PublishRef,
+    /// The published attrpath
+    /// Should be fully resolved to avoid ambiguity
+    attr_path: AttrPath,
+    stability: Stability,
+    analysis: State,
+}
+
+impl<'flox> Publish<'flox, Empty> {
+    pub fn new(
+        flox: &'flox Flox,
+        publish_ref: PublishRef,
+        attr_path: AttrPath,
+        stability: Stability,
+    ) -> Publish<'flox, Empty> {
+        Self {
+            flox,
+            publish_ref,
+            attr_path,
+            stability,
+            analysis: Empty,
+        }
+    }
+
+    /// run analysis on the package and switch to next state
+    ///
+    /// It uses an analyzer flake to extract eval metadata of the derivation.
+    /// The analyzer applies a function to all packages in a `target` flake
+    /// and provides the result under `#analysis.eval.<full attrpath of the package>`.
+    ///
+    /// We evalaute this analysis as json, to which we add
+    /// * source urls for reproducibility
+    /// * the nixpkgs stability being used to create the package
+    pub async fn analyze(self) -> PublishResult<Publish<'flox, NixAnalysis>> {
+        let mut drv_metadata_json = self.get_drv_metadata().await?;
+        let flake_metadata = self.get_flake_metadata().await?;
+
+        // DEVIATION FROM BASH: using `locked` here instead of `resolved`
+        //                      this is used to reproduce the package,
+        //                      but is essentially redundant because of the `source.locked`
+        drv_metadata_json["element"]["url"] = json!(flake_metadata.locked.to_string());
+        drv_metadata_json["source"] = json!({
+            "locked": flake_metadata.locked,
+            "original": flake_metadata.original,
+            "remote": flake_metadata.original,
+        });
+        drv_metadata_json["eval"]["stability"] = json!(self.stability);
+
+        Ok(Publish {
+            flox: self.flox,
+            publish_ref: self.publish_ref,
+            attr_path: self.attr_path,
+            stability: self.stability,
+            analysis: NixAnalysis(drv_metadata_json),
+        })
+    }
```

Comment:
We may want to enforce a "build" state before analyzing, after all a publish of a package that cant be built is not valid anyway.

Reply:
hmm, when Publish is run from the cli impl, we can expect that the flake and attr_path exist.

on checking:

- `build` would likely be the most expensive step (assuming you did not run `flox build` immediately before)
- a package that does not build does not need to be published
- a package that is already published (same version, same hash) does not need to be published again, this can only be checked after evaluating the outpaths though..

With that, the `built` state would come _after_ the `NixAnalysis`
I' not sure how finely you want to implement these states..